### ISSUE TITLE
NanoRT struct not defined

### ISSUE THREAD

**Issue by syoyo** [jeffamstutz/module_nanort#1]
os: macosx 10.11
OSPRay : release-1.0(`master` branch)
ispc : v1.9.1

I have cloned `module_nanort` into OSPRay's module directory, but failed to compile.

```
[ 65%] Building CXX object builddir/modules/intel64/module_nanort/CMakeFiles/ospray_ospray_module_nanort.dir/NanoRT.cpp.o
/Users/syoyo/work/OSPRay/modules/module_nanort/NanoRT.cpp:70:36: error: unknown type name 'NanoRT'
static void nanortBoundsFunc(const NanoRT* nrts,
                                   ^
```

The definition of `NanoRT` struct is not found in ispc generated header file(`NanoRT_ispc.h`).
Probably we need to provide `NanoRT.ih` header file and implement `NanoRT` struct there,  or something changed in `develop` branch?

**Comment by jeffamstutz** [comment]
Ha, I know exactly what this is....I used "NanoRT.h" as the header for the OSPRay Geometry, which on OS X conflicts with the "nanort.h" file (I wrote this on a Linux machine). OS X doesn't respect casing in file names. Just fixed it, try again. :)

**Comment by jeffamstutz** [comment]
I'll leave this issue open until you've confirmed it works.

**Comment by syoyo** [comment]
Thanks for the fix, but there is still an error.

https://github.com/jeffamstutz/module_nanort/blob/master/NanoRT_Geom.cpp#L27

This should be `NanoRT_Geom_ispc.h`.

And also I got the following error.

```
[ 65%] Building CXX object builddir/modules/intel64/module_nanort/CMakeFiles/ospray_ospray_module_nanort.dir/NanoRT_Geom.cpp.o
/Users/syoyo/work/OSPRay/modules/module_nanort/NanoRT_Geom.cpp:134:40: error: unknown type name 'RTCIntersectContext'
                                 const RTCIntersectContext* context,
                                       ^
/Users/syoyo/work/OSPRay/modules/module_nanort/NanoRT_Geom.cpp:135:34: error: unknown type name 'RTCRayNp'
                                 RTCRayNp*                  rays,
                                 ^
```

`RTCIntersectContext` and other RTC structs are defined in recent embree(at least v2.11), but not defined in embree 2.7.1(which is used in OSPRay release 1.0).

**Comment by jeffamstutz** [comment]
Also an easy fix, just made a local copy of rtcore_ray.h that comes with the module which introduces missing components if the detected Embree version isn't high enough. Tested (and works) with packaged Embree 2.7.1 and 2.11.0 on my linux machine. We are in limbo with upgrading our internal Embree source in OSPRay, but this should get the module working for you anyway in the short-term.

Bah, sorry for incurring multiple "bumps in the road".

**Comment by syoyo** [comment]
I see. I changed to use external Embree(v2.11.0) and replaced `UNUSED` macro with `(void)` to fix compilation.

Also I create symlink to `libospray_ospray_module_nanort.dylib` as `libospray_module_nanort.dylib`

Then,

```
$ ./ospGlutViewer ~/data/objs/house.obj -m nanort --trianglemesh-type nanort
Loaded plugin 'nanort' ...
#osp:minisg:parseOBJ (Warning): cannot parse input line Wavefront OBJ material file
```

Although the program runs, but it does not use nanort plugin(it uses embree raytracer). Probably something is not working well for OSPRay plugin system?

**Comment by jeffamstutz** [comment]
Yeah, I should have put in the README that the module was built expecting our upstream development branch (`devel`) of OSPRay (UNUSED macro issue and why you had to make a symlink). Glad though that it at least builds now, and I assume runs OK. I'll update the README for others.

So the way I do the "ray forwarding" concept is through Embree's User Defined Geometry API (UDG).  The Embree UDG API takes 2 components: 1. a function which defines the bounds of the new "primitive", and 2. function(s) which intersect the new "primitive". Thus the new primitive I defined is an entire NanoRT BVH.

Thus what I did was implement a new triangle mesh OSPRay geometry, which keeps track of the bounding box for the entire geometry. Within the geometry class, a NanoRT BVH is built and used during ray intersection. When Embree encounters the bounding box for the NanoRT geometry and calls the "primitive" intersection function, all I do is translate that ray into a NanoRT ray and traverse the NanoRT BVH to get the intersection. (NOTE: I was super lazy about calculating the intersection geometric normal, as that comes with Embree). The reason the implementation is semi-complicated is that Embree UDG intersection functions must be defined based on Embree's packet size for our ISPC renderers. This means defining a 4-wide function on SSE, 8-wide AVX/AVX2, and 16-wide on KNL, which is seen in NanoRT_Geom.cpp:318-348. The "meat" of the module is the static `traceRay` function, which is on lines 86-125.

Embree DOES build a BVH over all instances of NanoRT geometries, which means that if ospNewGeometry("nanort") ends up getting called more than once, then there's essentially a 2-level BVH being created. I have some models which our loaders all throw into a single geometry, but ultimately this approach was used because it was trivial compared to making Embree a substitutable component inside OSPRay. I'm not saying that wouldn't be a good thing....it's just MUCH more work. :)
